ttf2font.1
.TH TTF2FONT
.SH NAME
TTF2font \- Utility program designed to create font bitmaps for Cube Engine games
.SH SYNOPSIS
.B ttf2font
[
.B \-?
]
[
.BI \-h homedir
]
[
.BI \-n outname
]
[
.BI \-i imgsize
]
[
.BI \-s fonsize
]
[
.BI \-p padsize
]
[
.BI \-d shdsize
]
[
.BI \-o {0|1}
]
[
.BI \-g {0..3}
]
[
.BI \-q {0..2}
]
[
.BI \-c {0..9}
]
.BI \-f fontfile
|
.I fontfile
.SH DESCRIPTION
TTF2Font is a utility program designed to create font bitmaps for Cube Engine games, it works by taking a Truetype font and building it into a set of coordinates in an image. It was created by Quinton "quin" Reeves and Lee "eihrul" Salzman, to assist Joshua L. "verbalshadow" Blocher in his creation of content for Blood Frontier.
.PP
The program will create two files upon proper execution, based on the name of the font you provide it. Should you wish to use this as your default font in game, you will need to rename the config file "font.cfg" (and optionally, rename the image "font.png") and then update the config file accordingly. The font will not work in the game as output by the program, you must edit it.
.SH OPTIONS
.TP
.B \-?
Shows the usage help for the program.
.TP
.BI -h homedir
Sets the home directory to
.IR homedir ,
defaults to the current directory ("."). This is an alternative location in which to look for files, and also output the files the program creates.
.TP
.BI \-n outname
Provides the name of the output as
.IR outname ,
defaults to "default". The file names, and "font name" will be named with this.
.TP
.BI \-i imgsize
Sets the initial dimensions for the generated image to
.IR imgsize ,
defaults to "512". If the program exceeds this size it will increase this value until it is able to produce a working image. Images may only be powers of two in size, if it is not, the program will adjust it automatically.
.TP
.BI \-s fonsize
Defines the point size of the font to use to
.IR fonsize ,
defaults to "56".
.TP
.BI \-p padsize
Adds
.I padsize
pixels of padding around each character, defaults to "1".  This is useful if pixels "bleed" into the characters around it. The generated dimensions of each character will not include this space when displayed in game.
.TP
.BI \-d shdsize
Create a shadow behind each character with an offset of
.IR shadowoffset ,
defaults to "2". This helps make characters more readable when the background is bright, while also creating a nice depth effect.
.TP
.BI \-o {0|1}
Use an outline around the entire character for a shadow if set to "1", defaults to "1". This uses the depth shadow argument to determine its size.
.TP
.BI \-g {0..3}
Sets the game which you are making the font for, defaults to "0" (no game). Valid values are:
.IP
   * 0: No game, neutral file output which must be edited
   * 1: Sauerbraten (and derivatives)
   * 2: AssaultCube
   * 3: Blood Frontier/Red Eclipse
.TP
.BI \-q {0..2}
Makes Truetype fonts render with specified quality, defaults to "2". A value of "2" will antialias characters it renders, "1" will do a simple blend on them, while "0" will not enhance them at all.
.TP
.BI \-c {0..9}
Compresses the generated PNG image at given level, defaults to "1" (or "Best Speed"). Higher values will compress the image more at the expense of longer loading times.
.TP
.BI \-f fontfile " | " fontfile
Specifies the path to the Truetype font to use. This argument is required for proper operation.
.SH EXAMPLES
.B ttf2font -nmyfontbitmap myfont.ttf -d10 -o0 -g3
.RS 4
Creates a font bitmap with the name and filename "myfontbitmap" from the original font myfont.ttf, with a distinct drop shadow and a configfile setup to be compatible with Blood Frontier or Red Eclipse.
.RE
.SH AUTHORS
TTF2font was created by Quinton "quin" Reeves and Lee "eihrul" Salzman.
.PP
This manpage was adapted from the included readme.txt by Martin Erik Werner <martinerikwerner@gmail.com>
.SH HISTORY
v0.1
    * initial release

v0.2
    * alpha blending fixes
    * -w arg to supress win32 msg boxes

v0.3
    * -o arg to use an outline around the glyph
    * -g arg to set output game format
    * -n arg to set the name of the output


src/shared/ttf2font.cpp
//
// ttf2font: make font bitmap and cfg from truetype fonts
// (C) 2008 Blood Frontier - ZLIB License
//
// Created by Quinton "quin" Reeves and Lee "eihrul" Salzman
//
// Usage:
//
// 	-?			this help
// 	-h<S>		home dir, output directory
// 	-f<S>		font file
// 	-n<S>		output name
// 	-g<N[0..3]>	game type
// 	-i<N>		image size
// 	-s<N>		font size
// 	-p<N>		char padding
// 	-d<N>		shadow depth
// 	-o<N>		outward shadow
// 	-q<N[0..2]>	render quality
// 	-c<N[0..9]>	compress level
//
#include "pch.h"
#include "tools.h"
#include <errno.h>

#ifdef main
#undef main
#endif

#if SDL_BYTEORDER == SDL_BIG_ENDIAN
#define RMASK 0xff000000
#define GMASK 0x00ff0000
#define BMASK 0x0000ff00
#define AMASK 0x000000ff
#else
#define RMASK 0x000000ff
#define GMASK 0x0000ff00
#define BMASK 0x00ff0000
#define AMASK 0xff000000
#endif

enum { GAME_NONE = 0, GAME_SAUER, GAME_AC, GAME_BF, GAME_MAX };

static struct gametypes
{
	const char *fontdir;
} game[] = {
	{ ""				},
	{ "data/"			},
	{ "packages/misc/"	},
	{ "fonts/"			},
};

const char *program, *fonname = NULL, *outname = "default";
string cutname, imgname, cfgname;
int retcode = -1, win32msg = 0, gametype = GAME_NONE,
	imgsize = 0, fonsize = 0, padsize = 1, shdsize = 2, outline = 1,
		quality = 2, pngcomp = Z_BEST_SPEED;

void conoutf(const char *text, ...)
{
	string str;
	va_list vl;

	va_start(vl, text);
	vsnprintf(str, _MAXDEFSTR-1, text, vl);
	va_end(vl);

#ifdef WIN32
	if(win32msg)
		MessageBox(NULL, str, "TTF2Font", MB_OK|MB_ICONINFORMATION|MB_TOPMOST);
	else
#endif
		fprintf(stdout, "%s\n", str);
}

void erroutf(const char *text, ...)
{
	string str;
	va_list vl;

	va_start(vl, text);
	vsnprintf(str, _MAXDEFSTR-1, text, vl);
	va_end(vl);

#ifdef WIN32
	if(win32msg)
		MessageBox(NULL, str, "TTF2Font Error", MB_OK|MB_ICONERROR|MB_TOPMOST);
	else
#endif
		fprintf(stderr, "%s\n", str);
}

void savepng(SDL_Surface *s, const char *name)
{
	FILE *p = openfile(name, "wb");
	if(p)
	{
		png_structp g = png_create_write_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
		if(g)
		{
			png_infop i = png_create_info_struct(g);
			if(i)
			{
				if(!setjmp(png_jmpbuf(g)))
				{
					png_init_io(g, p);

					png_set_compression_level(g, pngcomp);

					png_set_IHDR(g, i, s->w, s->h, 8,
						PNG_COLOR_TYPE_RGB_ALPHA, PNG_INTERLACE_NONE,
							PNG_COMPRESSION_TYPE_DEFAULT, PNG_FILTER_TYPE_DEFAULT);

					png_write_info(g, i);

					uchar *d =(uchar *)s->pixels;
					loopk(s->h)
					{
						png_write_row(g, d);
						d += s->w*s->format->BytesPerPixel;
					}
					png_write_end(g, NULL);
				}
			}
			else erroutf("Failed to create png info struct.");
			png_destroy_write_struct(&g, &i);
		}
		else erroutf("Failed to create png write struct.");
		fclose(p);
	}
	else erroutf("Failed to open %s: %s", name, strerror(errno));
}

// only works on 32 bit surfaces with alpha in 4th byte!
void blitchar(SDL_Surface *dst, SDL_Surface *src, int x, int y)
{
    uchar *dstp = (uchar *)dst->pixels + y*dst->pitch + x*4,
          *srcp = (uchar *)src->pixels;
    int dstpitch = dst->pitch - 4*src->w,
        srcpitch = src->pitch - 4*src->w;
    loop(dy, src->h)
    {
        loop(dx, src->w)
        {
            uint k1 = (255U - srcp[3]) * dstp[3], k2 = srcp[3] * 255U, kmax = max(dstp[3], srcp[3]), kscale = max(kmax * 255U, 1U);
            dstp[0] = (dstp[0]*k1 + srcp[0]*k2) / kscale;
            dstp[1] = (dstp[1]*k1 + srcp[1]*k2) / kscale;
            dstp[2] = (dstp[2]*k1 + srcp[2]*k2) / kscale;
            dstp[3] = kmax;
            dstp += 4;
            srcp += 4;
        }
        dstp += dstpitch;
        srcp += srcpitch;
    }
}

enum { CF_NONE = 0, CF_OK, CF_FAIL, CF_SIZE };

#define FONTSTART	33
#define FONTCHARS	94

struct fontchar { char c; int x, y, w, h; };

int tryfont(int isize, int fsize, bool commit)
{
	int cf = CF_NONE;

	TTF_Font *f = TTF_OpenFont(findfile(fonname, "r"), fsize);

	if(f)
	{
		if(!win32msg && !commit)
			conoutf("Attempting to build size %d font in dimensions %dx%d...", fsize, isize, isize);

		SDL_Surface *t = SDL_CreateRGBSurface(SDL_SWSURFACE, isize, isize, 32, RMASK, GMASK, BMASK, AMASK);
		if(t)
		{
			if(commit)
			{
				loopi(t->h) memset((uchar *)t->pixels + i*t->pitch, 0, 4*t->w);
			}

			vector<fontchar> chars;
			SDL_Color c[3] = { { 1, 1, 1, }, { 255, 255, 255 }, { 0, 0, 0 } };
			int x = padsize, y = padsize, h = 0, ma = 0, mw = 0, mh = 0;

			loopi(FONTCHARS)
			{
				fontchar &a = chars.add();
				a.c = i+FONTSTART;

				s_sprintfd(m)("%c", a.c);
				SDL_Surface *s[2] = { NULL, NULL };
				loopk(commit ? 2 : 1)
				{
					switch(quality)
					{
						case 0:
						{
							s[k] = TTF_RenderText_Solid(f, m, c[k]);
							break;
						}
						case 1:
						{
							s[k] = TTF_RenderText_Shaded(f, m, c[k], c[2]);
							if(s[k])
								SDL_SetColorKey(s[k], SDL_SRCCOLORKEY,
									SDL_MapRGBA(s[k]->format, c[2].r, c[2].g, c[2].b, 0));
							break;
						}
						case 2:
						default:
						{
							s[k] = TTF_RenderText_Blended(f, m, c[k]);
							if(s[k]) SDL_SetAlpha(s[k], 0, 0);
							break;
						}
					}
					if(s[k])
					{
						SDL_Surface *c = SDL_ConvertSurface(s[k], t->format, SDL_SWSURFACE);
						SDL_FreeSurface(s[k]);
						s[k] = c;
					}
					if(!s[k])
					{
						if(k && s[0]) SDL_FreeSurface(s[0]);
						erroutf("Failed to render font: %s", TTF_GetError());
						cf = CF_FAIL;
						break;
					}
				}

				if(!cf)
				{
					int osize = (shdsize*(outline ? 2 : 1)), nsize = outline ? shdsize : 0;
					a.x = x;
					a.y = y;
					a.w = s[0]->w+osize;
					a.h = s[0]->h+osize;

					if(a.x+a.w >= t->w)
					{
						a.y = y = a.y+h+padsize;
						a.x = x = h = 0;
					}

					if(a.y+a.h >= t->h)
					{
						if(!win32msg)
							conoutf("Size %d font exceeded dimensions %dx%d at %d.", fsize, t->w, t->h, a.y+a.h);
						cf = CF_SIZE; // keep going, we want optimal sizes
						break;
					}

					if(commit)
					{
						loopk(2) if(k || shdsize)
						{
							bool sd = shdsize && !k;
							if(sd && outline)
							{
								loop(dx, osize+1)
								{
									loop(dy, osize+1)
									{
										blitchar(t, s[k], a.x+dx, a.y+dy);
									}
								}
							}
							else blitchar(t, s[k], a.x+(sd?osize:nsize), a.y+(sd?osize:nsize));
						}
					}

					x += a.w+padsize;
					if(a.h > h) h = a.h;
					mw += a.w;
					mh += a.h;
					ma += (a.w+padsize)*(a.h+padsize);
				}
				else break;
			}

			if(!cf)
			{
				if(commit)
				{
					FILE *p = openfile(cfgname, "w");
					if(p)
					{
						int ow = mw/FONTCHARS, oh = mh/FONTCHARS;

						savepng(t, imgname);

						fprintf(p, "// font generated by ttf2font\n\n");
						fprintf(p, "font %s \"%s%s\" %d %d\n\n", outname, game[gametype].fontdir, gametype != GAME_BF ? imgname : outname, ow, oh);
						loopv(chars)
						{
							if(chars[i].h != oh)
							{
								fprintf(p, "fontchar\t%d\t%d\t%d\t%d\t\t// %c\n",
									chars[i].x, chars[i].y, chars[i].w, chars[i].h, chars[i].c);
							}
							else if(chars[i].w != ow)
							{
								fprintf(p, "fontchar\t%d\t%d\t%d\t\t\t// %c\n",
									chars[i].x, chars[i].y, chars[i].w, chars[i].c);
							}
							else
							{
								fprintf(p, "fontchar\t%d\t%d\t\t\t\t// %c\n",
									chars[i].x, chars[i].y, chars[i].c);
							}
						}
						fclose(p);

						if(!win32msg) conoutf(" "); // seperate the previous output nicely

						conoutf("Completed conversion of font at size %d, dimensions %dx%d.\nUsed %d of %d surface pixels (%.1f%%).\nNOTE: You must edit %s, it does not come ready to use.",
							fsize, t->w, t->h,
							ma, t->w*t->h, (float(ma)/float(t->w*t->h))*100.f,
							cfgname
						);
						cf = CF_OK;
					}
					else erroutf("Failed to open %s: %s", cfgname, strerror(errno));
				}
				else cf = CF_OK;
			}
			chars.setsize(0);
			SDL_FreeSurface(t);
		}
		else erroutf("Failed to create SDL surface: %s", SDL_GetError());

		TTF_CloseFont(f);
	}
	else erroutf("Failed to open font: %s", TTF_GetError());

	return cf ? cf : CF_FAIL; // CF_NONE is only used internally..
}

void makefont()
{
	int isize = imgsize ? imgsize : 512, fsize = fonsize ? fonsize : (imgsize ? 256 : 64);
	bool commit = false;

	while(retcode < 0)
	{
		switch(tryfont(isize, fsize, commit))
		{
			case CF_OK:
			{
				if(commit) retcode = EXIT_SUCCESS;
				else commit = true;
				break;
			}
			case CF_FAIL:
			{
				retcode = EXIT_FAILURE;
				break;
			}
			case CF_SIZE:
			default:
			{
				if(!fonsize || imgsize) fsize--;
				else if(!imgsize || fonsize) isize += 2;
				else
				{
					erroutf("Failed to create the size %d font in the dimensions %dx%d.", fsize, isize);
					retcode = EXIT_FAILURE;
				}
				break; // continue trying then..
			}
		}
	}
}

void usage()
{
	conoutf("%s usage:\n\n\t-?\t\tthis help\n\t-h<S>\t\thome dir, output directory\n\t-g<N[0..3]>\tgame type\n\t-f<S>\t\tfont file\n\t-n<S>\t\toutput name\n\t-i<N>\t\timage size\n\t-s<N>\t\tfont size\n\t-p<N>\t\tchar padding\n\t-d<N>\t\tshadow depth\n\t-o<N[0..1]>\toutward shadow\n\t-q<N[0..2]>\trender quality\n\t-c<N[0..9]>\tcompress level", program);
}

int main(int argc, char *argv[])
{
	program = argv[0];
	sethomedir(".");

	for(int i = 1; i < argc; i++)
	{
		if(argv[i][0]=='-') switch(argv[i][1])
		{
			case '?': retcode = EXIT_SUCCESS; break;
#ifdef WIN32
			case 'w': win32msg = clamp(atoi(&argv[i][2]), 0, 1); break;
#endif
			case 'h': sethomedir(&argv[i][2]); break;
			case 'f': fonname = &argv[i][2]; break;
			case 'n': outname = &argv[i][2]; break;
			case 'g': gametype = clamp(atoi(&argv[i][2]), int(GAME_NONE), int(GAME_MAX-1)); break;
			case 'i': imgsize = max(atoi(&argv[i][2]), 0); imgsize -= imgsize%2; break;
			case 's': fonsize = max(atoi(&argv[i][2]), 0); break;
			case 'p': padsize = max(atoi(&argv[i][2]), 0); break;
			case 'd': shdsize = max(atoi(&argv[i][2]), 0); break;
			case 'o': outline = clamp(atoi(&argv[i][2]), 0, 1); break;
			case 'q': quality = clamp(atoi(&argv[i][2]), 0, 2); break;
			case 'c': pngcomp = clamp(atoi(&argv[i][2]), Z_NO_COMPRESSION, Z_BEST_COMPRESSION); break;
			default:
			{
				erroutf("Unknown command line argument -%c", argv[i][1]);
				retcode = EXIT_FAILURE;
				break;
			}
		}
		else fonname = argv[i]; // for drag and drop on win32. or laziness
	}

	if(retcode < 0 && fonname)
	{
		const char *dot = strpbrk(fonname, ".");
		if(dot) s_strncpy(cutname, fonname, dot-fonname+1);
		else s_strcpy(cutname, fonname);

		if(!*outname) outname = cutname;

		s_sprintf(imgname)("%s.png", outname);
		s_sprintf(cfgname)("%s.cfg", outname);

		if(SDL_Init(SDL_INIT_NOPARACHUTE) != -1)
		{
			if(TTF_Init() != -1)
			{
				makefont();
				TTF_Quit();
			}
			else erroutf("Failed to initialize TTF: %s", TTF_GetError());

			SDL_Quit();
		}
		else erroutf("Failed to initialize SDL: %s", SDL_GetError());
	}
	else usage();

	return retcode < 0 ? EXIT_FAILURE : retcode;
}
